pp.c

```diff
@@ -3717,6 +3717,112 @@ PP_wrapped(pp_substr,
     RETPUSHUNDEF;
 }
 
+/* OP_SUBSTR_LEFT is a specialized version of OP_SUBSTR, where:
+ *     the EXPR is a PADSV
+ *     the OFFSET is a CONST zero
+ *     the replacement pattern is a CONST ""
+ *     it's definitely not in lvalue context (see the check in pp_substr)
+ *     it definitely doesn't have OPpSUBSTR_REPL_FIRST set
+ *     it may be an rvalue or in void context (may support TARGMY later)
+ */
+PP(pp_substr_left)
+{
+    dTARGET;
+    STRLEN curlen;
+    STRLEN utf8_curlen = 0;
+    STRLEN byte_len = 0;
+    SV *sv = PL_stack_sp[-1];
+    const bool rvalue = (GIMME_V != G_VOID) || (PL_op->op_private & OPpTARGET_MY);
+    const bool do_chop = (MAXARG3 == 4);
+    const char *tmps;
+
+    if (SvROK(sv) && do_chop) {
+        Perl_ck_warner(aTHX_ packWARN(WARN_SUBSTR),
+                        "Attempt to use reference as lvalue in substr"
+        );
+    }
+
+    if (do_chop) {
+        SvGETMAGIC(sv);
+        if (!SvTHINKFIRST(sv) && SvPOK(sv)) {
+            curlen = SvCUR(sv);
+            tmps   = SvPVX_mutable(sv);
+            (void)SvPOK_only_UTF8(sv);
+        } else {
+            tmps = SvPV_force_nomg(sv, curlen);
+        }
+    } else {
+        if (!SvGMAGICAL(sv) && !SvROK(sv) && SvPOK(sv)) {
+            curlen = SvCUR(sv);
+            tmps = (char *)SvPVX_const(sv);
+        } else {
+            tmps = SvPV_const(sv, curlen);
```

Comment:
This seems unnecessary - why not just `tmps = SvPV_const(sv, curlen);`?

Under all the complexity this is all SvPV_const() does, except it doesn't do the unneeded SvROK() check and does the flag check in a single comparison.

Similarly for the SvPV_force_nomg unwrapping above that.

Reply:
Oh, yes, thanks. I had a brain fart about Perl_SvPV_helper, figured it out, but forgot to undo this unwrapping.